src/finitediff.jl

```diff
@@ -14,266 +14,41 @@ end
-    elseif fdtype == Val{:complex}
-        epsilon = eps(T)
-        @. df = imag(f(x+im*epsilon)) / epsilon
-    end
-    df
-end
-
-function finite_difference!(df::AbstractArray{T}, f, x::T, fdtype::DataType, fx::AbstractArray{T}, ::Type{Val{:DiffEqDerivativeWrapper}}) where T<:Real
-    fx1 = f.fx1
-    if fdtype == Val{:forward}
-        epsilon = compute_epsilon(fdtype, x)
-        f(fx, x)
-        f(fx1, x+epsilon)
-        @. df = (fx1 - fx) / epsilon
-    elseif fdtype == Val{:central}
-        epsilon = compute_epsilon(fdtype, x)
-        f(fx, x-epsilon)
-        f(fx1, x+epsilon)
-        @. df = (fx1 - fx) / (2 * epsilon)
-    elseif fdtype == Val{:complex}
-        epsilon = eps(T)
-        f(fx, f(x+im*epsilon))
-        @. df = imag(fx) / epsilon
-    end
-    df
-end
-
-#=
-Compute the derivative df of a real-valued callable f on a collection of points x.
-Optimized implementations for StridedArrays.
-=#
-function finite_difference!(df::StridedArray{T}, f, x::StridedArray{T}, ::Type{Val{:central}}, ::Union{Void,StridedArray{T}}, ::Type{Val{:Default}}) where T<:Real
-    epsilon_factor = compute_epsilon_factor(Val{:central}, T)
-    @inbounds for i in 1 : length(x)
-        epsilon = compute_epsilon(Val{:central}, x[i], epsilon_factor)
-        epsilon_double_inv = one(T) / (2*epsilon)
-        x_plus, x_minus = x[i]+epsilon, x[i]-epsilon
-        df[i] = (f(x_plus) - f(x_minus)) * epsilon_double_inv
-    end
-    df
-end
-
-function finite_difference!(df::StridedArray{T}, f, x::StridedArray{T}, ::Type{Val{:forward}}, ::Void, ::Type{Val{:Default}}) where T<:Real
-    epsilon_factor = compute_epsilon_factor(Val{:forward}, T)
-    @inbounds for i in 1 : length(x)
-        epsilon = compute_epsilon(Val{:forward}, x[i], epsilon_factor)
-        epsilon_inv = one(T) / epsilon
-        x_plus = x[i] + epsilon
-        df[i] = (f(x_plus) - f(x[i])) * epsilon_inv
-    end
-    df
-end
-
-function finite_difference!(df::StridedArray{T}, f, x::StridedArray{T}, ::Type{Val{:forward}}, fx::StridedArray{T}, ::Type{Val{:Default}}) where T<:Real
-    epsilon_factor = compute_epsilon_factor(Val{:forward}, T)
-    @inbounds for i in 1 : length(x)
-        epsilon = compute_epsilon(Val{:forward}, x[i], epsilon_factor)
-        epsilon_inv = one(T) / epsilon
-        x_plus = x[i] + epsilon
-        df[i] = (f(x_plus) - fx[i]) * epsilon_inv
-    end
-    df
-end
-
-#=
-Compute the derivative df of a real-valued callable f on a collection of points x.
-Single point implementations.
-=#
-function finite_difference(f, x::T, fdtype::DataType, f_x::Union{Void,T}=nothing) where T<:Real
-    if fdtype == Val{:complex}
-        epsilon = eps(T)
-        return imag(f(x+im*epsilon)) / epsilon
+function compute_epsilon_elemtype(epsilon, x)
+    if typeof(epsilon) != Void
+        return eltype(epsilon)
+    elseif eltype(x) <: Real
+        return eltype(x)
+    elseif eltype(x) <: Complex
+        return eltype(x).parameters[1]
```

Comment:
is this type-inferrable?

Reply:
Probably better to:

```julia
julia> real(typeof(1.0+im))
Float64
```

Reply:
Good point, changed it to `real(eltype(x))`.